### DISCUSSION TITLE
ESP32S3 USB MIDI problem

### DISCUSSION THREAD

**Discussion by aberoev** [tttapa/Control-Surface#1004] [General]
Hello 

Thank you for the great library I have made a DAW controller using Arduino pro micro, love it!!

I am now trying to make ESP32S3 board work, but with no success. I installed espressifs arduino core v 3 [like suggested](https://tttapa.github.io/Control-Surface-doc/Doxygen/d8/d4a/md_pages_MIDI-over-USB.html#espressif-esp32-boards-with-usb-support) in the documentation and chose ESP32S3_USB-OTG as board

[This example sketch](https://github.com/tttapa/Control-Surface/blob/main/examples/1.%20MIDI%20Output/3.%20Rotary%20Encoders/AbsoluteRotaryEncoder/AbsoluteRotaryEncoder.ino) compiles (although incredibly slow, takes a couple of minutes!) and uploads, but the board does not show up as a USB MIDI device on my PC. 

The board is a[ Lilygo T-QT Pro](https://github.com/Xinyuan-LilyGO/T-QT). 

As for the very slow compiling issue, I have tried installing IDE 1 as suggested by some users with the same problem, but it is still very very slow on my windows machine. Makes trouble shooting this issue very time consuming :)))

Any hints? Thank you

**Comment by tttapa** [comment]
> chose ESP32S3_USB-OTG as board

Are you sure this is the correct board? The link you posted seems to recommend using the `ESP32S3 Dev Module`, which is also the one I used during testing. The ESP32S3_USB-OTG is quite a special board.

Did you select `USB-OTG (TinyUSB)` from the `Tools > USB Mode` menu?

Do you see any warnings during compilation?

> although incredibly slow, takes a couple of minutes!

A cold build takes around 1 minute on my 2017 i7-7700HQ (running Ubuntu 22.04), and a re-build takes around 25 seconds. This is quite a bit slower than some of the 8-bit AVR Arduinos, but not too bad considering the difference in features between an ESP32 and an AVR.

**Reply by aberoev** [reply]
Hi tttapa thank you for your reply, I think you are right I misunderstood which (board) core to use. I tried your suggestions but still with no luck

These are my settings

![esp32s3settings](https://github.com/tttapa/Control-Surface/assets/155575523/9e5882ec-a570-42a6-94f9-060c2920bc54)

I would really love to have this little board working as the included screen, wireless functions and overall thumb size is very interesting!!

There are no errors when compiling, and example sketches from the lilygo t-qt library uploads and works fine

**Reply by tttapa** [reply]
Try adding this at the bottom of your sketch. It should show a compilation error, please post it here.

```cpp
template <class> struct PrintType;
PrintType<decltype(midi.backend)> print_type;
```

**Reply by aberoev** [reply]
This is the compilation errors error I get with this sketch



>  .....\sketch_feb8a\sketch_feb8a.ino:29:35: error: aggregate 'PrintType<cs::ESP32_USBDeviceMIDIBackend> print_type' has incomplete type and cannot be defined
>    29 | PrintType<decltype(midi.backend)> print_type;
>       |                                   ^~~~~~~~~~
> 
> exit status 1
> 
> Compilation error: aggregate 'PrintType<cs::ESP32_USBDeviceMIDIBackend> print_type' has incomplete type and cannot be defined
>

**Reply by tttapa** [reply]
That looks alright. How are you verifying that MIDI over USB is not working?

Does the [arduino-esp32 example](https://github.com/espressif/arduino-esp32/blob/master/libraries/USB/examples/MIDI/MidiMusicBox/MidiMusicBox.ino) work for you?

**Reply by aberoev** [reply]
No that does not work either. I am using midimonitor.com and Steinberg Cubase to to see connected usb midi devices. I have tried unplugging, rebooting pc, making sure com port is not in use by other applications etc. When I plug in my Arduino Pro micro project it presents itself as expected. This is the chip 

![unnamed (12)](https://github.com/tttapa/Control-Surface/assets/155575523/31dff7b1-bce6-4515-aa94-a7588717e2e2)

**Reply by tttapa** [reply]
I'm not too familiar with Windows, but you may be running into this issue: https://stackoverflow.com/questions/52946832/re-enumeration-of-usb-device-which-is-upgraded-from-old-to-new-firmware-with-cha

**Reply by aberoev** [reply]
tttapa you are right! I tried connecting the esp32 to my wifes macintosh and lo and behold, the thing shows up. The [arduino-esp32 example ](https://github.com/espressif/arduino-esp32/blob/master/libraries/USB/examples/MIDI/MidiMusicBox/MidiMusicBox.ino) presented not one but five usb midi instances, and the control-surface example sketch presented one device.

This is great news, however my wife needs her computer back, so I will have to investigate how to fix this issue with my windows pc. I tried deleting the driver in device manager, and replugging the device, still nothing! The forum thread you linked to suggested changing iSerial in the USB descriptor - as far as I understand I will have to find out how to reconfigure the actual usb stack "tinyusb" - which is not directly exposed in control-surface library is that correct?

**Reply by tttapa** [reply]
Correct. I haven't looked into this in detail, but I believe you should be able to add `USB.serialNumber("foobar")` at the beginning of your `setup` function to change the serial number for the USB descriptors.

**Reply by aberoev** [reply]
tttapa I tried adding` USB.serialNumber("00000095")` (or another random number) to my setup of both the arduinio-esp32 example and the simple control-surface example with no luck. I tried uninstalling device drivers and rebooting my computer, tried different usb ports, reflashing the unit manually entering bootloader mode, etc but nothing will make the damn thing appear as a midi device on my windows pc (but still shows fine on the mac)

I stumbled upon [this comment](https://github.com/adafruit/Adafruit_TinyUSB_Arduino/issues/279#issuecomment-1527266012) and tried uploading the example [sketch](https://github.com/chegewara/EspTinyUSB/blob/master/examples/device/midi/midi.ino) - a midi device appeared on my windows pc!

I have insufficient knowledge of low level usb stuff to debug from here sadly

**Reply by tttapa** [reply]
You may want to try changing the PID:

```cpp
void setup() {
  USB.PID(USB.PID() ^ 0x0002);
  // ...
}
```

**Reply by aberoev** [reply]
Thank you for your suggestions tttapa, I tried adding that line but no dice. The unit still presents it self fine on my wifes macintosh, but not on my Windows PC. Out of curiosity I tried uploading a control surface sketch to a Raspberry Pi Pico board I had laying, using adafruit TinyUSB stack and earlephilhowers arduino pico core, and it works as expected, as does my atmega32u4 board. Unfortunately I don't have any other ESP32 boards than this Lilygo [T-QT Pro](https://github.com/Xinyuan-LilyGO/T-QT) to test with, I wonder if my Windows PC needs a specific driver in order to recognize the ESP32 CDC created. (Addition: I tried one of the [arduino esp32 HID keyboard sketches](https://github.com/espressif/arduino-esp32/blob/master/libraries/USB/examples/Keyboard/KeyboardMessage/KeyboardMessage.ino) which works fine on my Windows PC. Strange indeed)

**Reply by tttapa** [reply]
I've tried this on an old Windows 10 machine, and I can't seem to reproduce the issue. Could you please post the sketch you're using? It would be helpful if you could also enable "verbose output" during compilation (in the IDE preferences) and post the FQBN it prints at the top as well.

**Reply by aberoev** [reply]
Hi tttapa

Here is the verbose output when flashing the ESP32S3
[verboseoutput_esp32.txt](https://github.com/tttapa/Control-Surface/files/14322429/verboseoutput_esp32.txt)

These are the board settings 
![esp32settings](https://github.com/tttapa/Control-Surface/assets/155575523/8fca372e-cfe5-46b9-9014-8e9ac970bb4f)

One thing I notice FQBN says USB mode: Default - should it read TinyUSB instead?

Thank you very much for your patience and help with this

I am using [this simple example sketch](https://github.com/tttapa/Control-Surface/blob/main/examples/1.%20MIDI%20Output/3.%20Rotary%20Encoders/AbsoluteRotaryEncoder/AbsoluteRotaryEncoder.ino) of control surface library to test with

I have not started building a sketch for my use yet, I am hoping to build a Cubase single fader controller and use the Lilygo screen to display chosen channel number hopefully via callbacks from the DAW (have yet to explore this part, but having my computer / cubase recognize the board must be step number one lol)

**Reply by tttapa** [reply]
Thanks! I think the "Default" USB mode is correct, although the name is a bit misleading.

I've tried the example sketch you linked to with your exact settings on two different Windows 10 machines now, but I haven't been able to reproduce the issue. On both machines, the ESP32S3 shows up as a MIDI device in the Windows Device Manager, and I'm able to receive its MIDI messages in a MIDI monitor (I used https://www.roxxxtar.com/apps/miditool/ in the Edge browser).

Does the ESP32S3 show up in Device Manager on your computer? Here's what it looks like for me:
![image](https://github.com/tttapa/Control-Surface/assets/21296729/8bc38a01-d4a8-436d-af16-b9e8db31b367)

To rule out any issues related to the encoder, you may want to try a simpler test sketch, e.g. [MIDI-Output.ino](https://tttapa.github.io/Control-Surface-doc/Doxygen/d8/d9b/MIDI-Output_8ino-example.html).

**Reply by aberoev** [reply]
Tttapa

I am now sure this problem can be isolated to my Windows computer. I tried another Windows 11 computer and the thing shows up.

On my Windows 10 computer, there is no ESP32S3 midi device under Device manager. When I plug in my RP2040 pico test board, it shows midi instances just fine, as does a Pro Micro.

With the ESP32S3 I do see a USB serial device corresponding to the COM port shown in Arduino IDE, but no midi instance under Sound, video etc. 

I have no idea what this can be. Looks like a driver problem but to which piece of hardware?? Not too fond of wiping my Windows installation and starting over, but I might have to!

**Reply by aberoev** [reply]
Wait a minute, 
![image](https://github.com/tttapa/Control-Surface/assets/155575523/78e2bb7f-adb0-4037-8bd8-e049a5620819)

The device is shown in device manager as "Universal serial bus device" - not under Sound, video etc. It does not show up in morningstar midimonitor or Cubase as a midi device. So close yet so far :)

**Reply by tttapa** [reply]
I've never come across this issue before, and I don't use Windows that often, so I'm afraid I don't have any concrete suggestions. However, Microsoft does have some tools to inspect USB devices and their USB device descriptors etc.: https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/usbview

**Reply by aberoev** [reply]
Thank you tttapa

I will post to this thread if I find a solution. 

It seems the discover-able device presents itself as a USB JTAG debug unit
![image](https://github.com/tttapa/Control-Surface/assets/155575523/04c0d13b-4cfb-4ad7-99e4-fbb9e4bdff8e)

Under events, some interesting details appear:

![image](https://github.com/tttapa/Control-Surface/assets/155575523/bf0206d2-2fd8-420e-8e09-8262c30bb05e)

Device not migrated information:

`Device USB\VID_303A&PID_1001&MI_02\7&3371e176&0&0002 was not migrated due to partial or ambiguous match.

Last Device Instance Id: USB\VID_0451&PID_82EE\7&6f5980e&0&6
Class Guid: {88bae032-5a81-49f0-bc3d-a4ff138216d6}
Location Path: 
Migration Rank: 0xF000FFFFFFFFF122
Present: false
Status: 0xC0000719
`

**Reply by tttapa** [reply]
Since you're mentioning JTAG: GPIO3 (a pin used in the RotaryEncoder example) determines the JTAG source of the ESP32S3. I don't know if this is relevant, but I thought I'd mention it just in case. https://www.espressif.com/sites/default/files/documentation/esp32-s3-wroom-2_datasheet_en.pdf#page=15

**Reply by aberoev** [reply]
Hello tttapa just an update on this issue. I've tried updating to Windows 11, but the problem didn't go away. 

I just got a XIAO ESP32S3 module in the mail and immediately tried uploading a control surface sketch. It works beautifully on my PC. So apparently the Lilygo T-QT board just doesn't want to talk to my Windows computer with HID. Perhaps it is something with the arduino-esp32 core / tinyusb implementation for that specific board, I have no idea.

Both boards make a midi device appear on bespoke wife macbook, but only the Xiao board appears as a midi device on my Windows PC. 

When looking at a "Device Monitoring Studio" I get a midi instance with the Xiao board

![image](https://github.com/tttapa/Control-Surface/assets/155575523/05987a6c-5472-4ae6-9433-3558bf312724)

With the Lilygo board I get a USB bus device called TinyUSB Midi but it's not a midi instance.

![image](https://github.com/tttapa/Control-Surface/assets/155575523/ad903c3f-6fd9-461d-9f7c-5bdbc93fa4a0)


So, not a solution, but some more info for anyone interested.

**Comment by simonemarin** [comment]
Hallo I think I have the same issue with a ESP32-S3-Devkit v1.0 (ESP32-S3-WROOM-1). I flashed some very basic code just to get the interface instantiated but no midi device available, just as "USB JTAG/serial debug unit". Didn t check with other OS apart from Windows 11.

**Comment by simonemarin** [comment]
So, if anyone arrived here with the same issue: I had the current esp32 core definition in Arduino IDE (v2), so I changed it to the dev version as per instructions, it seemed like the board definition updated correctly but then I decided to reboot Windows as the usual last remedy and effectively when I started the IDE the boards were updated and now the Midi device si showig up correctly.
